pkg/monitor/azure/nsg/nsg.go

```diff
@@ -0,0 +1,176 @@
+package nsg
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"net/http"
+	"net/netip"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/arm"
+	mgmtnetwork "github.com/Azure/azure-sdk-for-go/services/network/mgmt/2020-08-01/network"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/sirupsen/logrus"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/metrics"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/network"
+)
+
+const (
+	DimSubnet              = "subnet"
+	DimVnet                = "vNet"
+	DimResourceGroup       = "resourcegroup"
+	DimSubscriptionID      = "subscriptionid"
+	DimSecurityGroup       = "networksecuritygroup"
+	DimNSGRuleName         = "rulename"
+	DimNSGRuleSources      = "sources"
+	DimNSGRuleDestinations = "destinations"
+	DimNSGRuleDirection    = "direction"
+	DimNSGRulePriority     = "priority"
+	DimClusterResourceID   = "clusterresourceid"
+	DimLocation            = "location"
+	DimTenantID            = "tenantid"
+
+	MetricInvalidDenyRule         = "preconfigurednsg.invalid.denyrule"
+	MetricSubnetAcessResponseCode = "preconfiguredNSG.subnetaccess.responsecode"
+	MetricSubnetAccessForbidden   = "preconfigurednsg.subnetaccess.forbidden"
+	MetricUnsuccessfulFPCreation  = "preconfigurednsg.fpcreation.unsuccessful"
+	MetricNSGMonitoringTimedOut   = "preconfigurednsg.monitoring.timedout"
+
+	expandNSG = "NetworkSecurityGroup"
+)
+
+// NSGMonitor is responsible for performing NSG rule validations when preconfiguredNSG is enabled
+type NSGMonitor struct {
+	log     *logrus.Entry
+	emitter metrics.Emitter
+	oc      *api.OpenShiftCluster
+
+	subnetCli network.SubnetsClient
+	done      chan error
+}
+
+func (n *NSGMonitor) Done() <-chan error {
+	return n.done
+}
+
+func NewNSGMonitor(log *logrus.Entry, oc *api.OpenShiftCluster, subscriptionID string, environment *azureclient.AROEnvironment, cred autorest.Authorizer, emitter metrics.Emitter) *NSGMonitor {
+	return &NSGMonitor{
+		log:     log,
+		emitter: emitter,
+		oc:      oc,
+
```

Comment:
In the cluster monitor, we construct a set of dimmensions specific to the monitor itself.  I.e. 
resourceId, subscriptionId, etc.  This means that every emit gauge or metric within the monitor itself will inherit these dimmensions, and save us a few lines of code.

Reply:
https://github.com/Azure/ARO-RP/blob/22a6d6063c3c18ec3a1e39bcdc107c7b403ac9d5/pkg/monitor/cluster/cluster.go#L65-L69